interfaces/prompting/errors/errors.go

```diff
@@ -0,0 +1,242 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// The errors package defines common error types which are used across the
+// prompting subsystems, along with constructors for specific errors based on
+// those broader types.
+package errors
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+var (
+	// NotFound errors when a prompt or rule is not found
+	ErrPromptNotFound = errors.New("cannot find prompt with the given ID for the given user")
+	ErrRuleNotFound   = errors.New("cannot find rule with the given ID")
+	ErrRuleNotAllowed = errors.New("user not allowed to request the rule with the given ID")
+
+	// Unsupported value errors, which are defined as functions below
+	ErrInvalidOutcome       = invalidOutcomeError
+	ErrInvalidLifespan      = invalidLifespanError
+	ErrRuleLifespanSingle   = ruleLifespanSingleError
+	ErrInvalidInterface     = invalidInterfaceError
+	ErrInvalidPermissions   = invalidPermissionsError
+	ErrPermissionsListEmpty = permissionsListEmptyError
+	// Parse errors, which are defined as functions below
+	ErrInvalidDuration    = invalidDurationError
+	ErrInvalidExpiration  = invalidExpirationError
+	ErrInvalidPathPattern = invalidPathPatternError
+	// Validation errors, which should never be used directly apart from
+	// checking errors.Is(), and should otherwise always be wrapped in
+	// dedicated error types defined below.
+	ErrReplyNotMatchRequestedPath        = errors.New("path pattern in reply constraints does not match originally requested path")
+	ErrReplyNotMatchRequestedPermissions = errors.New("permissions in reply constraints do not include all requested permissions")
+	ErrRuleConflict                      = errors.New("a rule with conflicting path pattern and permission already exists in the rule database")
+
+	// Internal errors which are not handled specifically
+	ErrPromptsClosed      = errors.New("prompts backend has already been closed")
+	ErrRulesClosed        = errors.New("rules backend has already been closed")
+	ErrTooManyPrompts     = errors.New("cannot add new prompts, too many outstanding")
+	ErrRuleIDConflict     = errors.New("internal error: rule with conflicting ID already exists in the rule database")
+	ErrRuleDBInconsistent = errors.New("internal error: interfaces requests rule database left inconsistent")
+
+	// Errors which are used internally and should never be returned over the API
+	ErrNoMatchingRule          = errors.New("no rule matches the given path")
+	ErrInvalidID               = errors.New("invalid ID: format must be parsable as uint64")
+	ErrRuleExpirationInThePast = errors.New("cannot have expiration time in the past")
+)
+
+// Unsupported value errors, which are built from the UnsupportedValueError struct
+
+func invalidOutcomeError(unsupported string, supported []string) *UnsupportedValueError {
```

Comment:
and rename the helpers to:
```suggestion
func NewInvalidOutcomeError(unsupported string, supported []string) *UnsupportedValueError {
```

so now it's more clear what really happens at the actual call site